### PR TITLE
fix(schema): correctly match namespace schema nodes, and correctly apply schema to new note when note existed as stub

### PR DESCRIPTION
**Pull request by hikchoi**
# fix(schema): correctly match namespace schema nodes, and correctly apply schema to new note when note existed as stub

This PR:
- Correctly matches notes with namespace schema nodes
  - namespace node match pattern was trimmed excessively and was not matching the node on the same depth as well as its immediate children
- Correctly apply schema when a note that existed as a stub is accepted as a new item through note lookup
  - When a note is created as a stub (because of a leaf note), it bypasses any schema application.
  - When this note is later accepted as a new note through lookup, this note that can potentially have a schema match does not get created with a schema applied.

### correctly matches namespace
1. create the following schema:
```
    version: 1
    imports: []
    schemas:
      - id: test
        children: 
          - testing
        title: test
        parent: root
      - id: testing
        pattern: "*"
        title: "testing"
        namespace: true
        template:
          id: template.test
          type: note
```
2. Create `template.test` with any note body
3. Create notes `test.test1` and `test.test1.test2` and see that `template.test` was applied to both of them
### correctly applies schema to newly accepted stub notes
1. Create the same schema describe above
2. Create note `test.one.two.three`
    - Note that this has no schema match, but it's parent that is a stub should match the schema above
3. Using lookup, create the note `test.one.two`.
    - This should already be in the quickpick as a stub
5. Since it has a matching schema, it should have the template applied to it.

# Pull Request Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [~] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
related: https://github.com/dendronhq/dendron/issues/3381

File: packages/common-all/src/dnode.ts [hunk 1]
```diff
@@ -1762,8 +1782,10 @@ export class SchemaUtils {
     const match = _.find(schemas, (sc) => {
       const pattern = SchemaUtils.getPatternRecursive(sc, schemaModule.schemas);
       if (sc?.data?.namespace && matchNamespace) {
+        // if we are matching a namespace node,
+        // match on the level of the node itself and its immediate children.
         namespace = true;
-        return minimatch(notePathClean, _.trimEnd(pattern, "/*"));
```
Comment (hikchoi) [inline packages/common-all/src/dnode.ts:1766]: The issue here was that `_.trimEnd` was removing the entire glob pattern when what we really wanted was to trim only one level of it.

**Comment by hikchoi** [comment]
Sorry about the scope creep with the second fix item. Found it while I was debugging the first issue and had to fix it for completeness 😄

File: packages/common-all/src/dnode.ts (hunk 1 above)
Comment (kevinslin) [reply packages/common-all/src/dnode.ts:1766]: its not immediately obvious what `slice(0,-2)` means - can you add some comments here?

Comment (hikchoi) [reply packages/common-all/src/dnode.ts:1766]: improved the comments to make clear what's happening here.

**Comment by hikchoi** [comment]
There was a breaking change with the engine method `writeNote` return signature change. rebased / fixed.
